File: day09/part2.py
import itertools

_TESTCASE = """\
35
20
15
25
47
40
62
55
65
95
102
117
150
182
127
219
299
277
309
576
""".strip().splitlines()


def compute_part1(data, size=25):
    for i, num in enumerate(data[size:], size):
        prev = data[i - size : i]
        if any(num == a + b for a, b in itertools.combinations(prev, 2)):
            continue
        else:
            break
    else:
        assert False

    return num


def compute(data, size=25):
    """
    >>> compute(_TESTCASE, 5)
    62
    """
    data = [*map(int, data)]
    num = compute_part1(data, size=size)

    start, stop = 0, 2

    while stop < len(data):
        span = data[start:stop]
        span_sum = sum(span)
        if span_sum < num:
            stop += 1
        elif span_sum > num:
            start += 1
        else:
            break
    else:
        assert False

    return min(span) + max(span)


def main():
    import pathlib

    input_path = pathlib.Path(__file__).with_name("input.txt")

    with input_path.open() as f:
        print(compute(f.read().splitlines()))


if __name__ == "__main__":
    main()

File: day09/part2.rs
use std::fs;

const SIZE: usize = 25;

fn compute(lines: Vec<&str>) -> i64 {
    let mut running_sum = Vec::new();
    let mut last_sum = 0;

    let numbers = lines
        .into_iter()
        .map(|s| s.parse().unwrap())
        .collect::<Vec<i64>>();

    let number = numbers
        .windows(SIZE + 1)
        .enumerate()
        .find_map(|(idx, window)| {
            let num = window[SIZE];

            if idx == SIZE + 1 {
                for n in &numbers[..=SIZE] {
                    last_sum += n;
                    running_sum.push(last_sum);
                }
            }

            last_sum += num;
            running_sum.push(last_sum);

            for i in 0..SIZE - 1 {
                for j in i + 1..SIZE {
                    if window[i] + window[j] == num {
                        return None;
                    }
                }
            }
            Some(num)
        })
        .expect("number not found");

    return running_sum
        .iter()
        .enumerate()
        .find_map(|(end_idx, cum_sum)| {
            if let Ok(start_idx) = running_sum.binary_search(&(cum_sum - number)) {
                let nums = &numbers[start_idx + 1..end_idx];
                return Some(nums.iter().min().unwrap() + nums.iter().max().unwrap());
            }
            None
        })
        .expect("no span found");
}

fn main() -> std::io::Result<()> {
    let path = fs::canonicalize(file!())?.with_file_name("input.txt");
    let file_content = fs::read_to_string(path)?;
    let lines: Vec<&str> = file_content.split_whitespace().collect();

    println!("{}", compute(lines));

    Ok(())
}
